### PR TITLE
HHH-11076 : test cases that show inconsistent collection initializati…

### PR DESCRIPTION
**Pull request by gbadner**
…on behavior

The fix causes some inconsistent results. See test cases for details.

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/jpa/test/hibernateFilters/ProxyPreservingFiltersOutsideInitialSessionTest.java [hunk 1]
```diff
@@ -71,4 +72,179 @@ public void testPreserveFilters() {
...
+					accountGroup.getAccounts().add( account );
+				}
+		);
+
+		doInJPA( this::entityManagerFactory, entityManager -> {
+					entityManager.unwrap( Session.class ).enableFilter( "byRegion" ).setParameter( "region", "US" );
+					AccountGroup accountGroup = entityManager.find( AccountGroup.class, 1L );
+					// Change the filter
+					entityManager.unwrap( Session.class ).enableFilter( "byRegion" ).setParameter( "region", "Europe" );
+					Hibernate.initialize( accountGroup.getAccounts() );
```
Comment (gbadner) [inline hibernate-core/src/test/java/org/hibernate/jpa/test/hibernateFilters/ProxyPreservingFiltersOutsideInitialSessionTest.java]: Compare this result with the result in testChangeFilterBeforeInitializeInTempSession.

File: hibernate-core/src/test/java/org/hibernate/jpa/test/hibernateFilters/ProxyPreservingFiltersOutsideInitialSessionTest.java [hunk 2]
```diff
@@ -71,4 +72,179 @@ public void testPreserveFilters() {
...
+
+		final AccountGroup group = doInJPA( this::entityManagerFactory, entityManager -> {
+					entityManager.unwrap( Session.class ).enableFilter( "byRegion" ).setParameter( "region", "US" );
+					return entityManager.find( AccountGroup.class, 1L );
+				} );
+
+		final AccountGroup mergedGroup = doInJPA( this::entityManagerFactory, entityManager -> {
+					return entityManager.merge( group );
+				} );
+
```
Comment (gbadner) [inline hibernate-core/src/test/java/org/hibernate/jpa/test/hibernateFilters/ProxyPreservingFiltersOutsideInitialSessionTest.java]: Compare the following behavior to what happens when the collection is initialized after Session#saveOrUpdate in testSaveOrUpdateNoFilterThenInitializeTempSession().

**Comment by vladmihalcea** [comment]
The `testChangeFilterBeforeInitializeInSameSession` can be fixed if the `AbstractPersistentCollection` does not create a copy of the filters.

So, instead of:

    enabledFilters = new HashMap<>( session.getLoadQueryInfluencers().getEnabledFilters() );

we could have:

    enabledFilters = session.getLoadQueryInfluencers().getEnabledFilters();

The second test that failed, `testSaveOrUpdateNoFilterThenInitializeTempSession` is debatable. First of all, only the Hibernate API offers the option to reattach entities, while JPA only offers merge. Because the entity is reattached, I guess that preserving filters should be the right option. While for `merge`, the filters should be reset.

**Comment by gbadner** [comment]
I'm still not convinced that the collection should "remember" the filter. I'll leave it to @sebersole to decide.

I looked at `testSaveOrUpdateNoFilterThenInitializeTempSession` and the assertion is already what you suggested.  Did you mean to say it that the expected value is 1?

**Comment by vladmihalcea** [comment]
No, it should be 3 if you change saveOrUpdate with merge.

Vlad

**Comment by vladmihalcea** [comment]
With saveOrUpdate, I believe we should retain the filters, but let's see
what Steve thinks of this.

**Comment by gbadner** [comment]
OK, I didn't realize your comment about the assertion related to changing from `saveOrUpdate` to `merge`. In any case, the statement already has an expected value of 3, because I find it surprising that the filter would be remembered in a new session.

**Comment by heruan** [comment]
@gbadner about your last concern, isn't the temporary session created ad-hoc just for that purpose? Then, it should have no filters enabled.

**Comment by gbadner** [comment]
@heruan, my concern is not with temporary sessions; it's with the impact of the change when the collection is attached to a non-temporary session.

**Comment by heruan** [comment]
I'm preparing a commit with several changes, like having the filters collected only if we have a `allow_lazy_load_no_trans` true, i.e. inside `prepareForPossibleLoadingOutsideTransaction()`. Would this be an acceptable trade-off? Or introducing a new setting maybe?

**Comment by beikov** [comment]
The test is already merged to master and annotated with `@FailureExpected` so I'll close this PR. Note that possible solutions to this problem are provided through the following PRs:

* https://github.com/hibernate/hibernate-orm/pull/3776
* https://github.com/hibernate/hibernate-orm/pull/3775

**Comment by sebersole** [comment]
I've said this on other discussions about this topic.  IMO we should not be saving these filters.  

As @gbadner points out, its not just about the loading outside of a transaction/Session.

The entire intent of this report is that the filters (really all LoadQueryInfluencers) that are in effect when the `PersistentCollection` is created are the filters that apply to that collection regardless of how it is used or when it is initialized.  E.g.

```
Session session = ...;

session.enableFilter( "abc" );

// Customer defines #orders as lazy
final Customer customer = session.load( Customer.class, 1 );

session.disableFilter( "abc" );

// what happens when we initialize orders?
customer.getOrders()...
```

**Pull request closed without merging**